### PR TITLE
string_view support, fmtlib usage and string/wstring support

### PR DESCRIPTION
**Pull request by flexferrum**

### CODE REVIEW

File: include/jinja2cpp/string_helpers.h [hunk 1]
```diff
@@ -0,0 +1,209 @@
+#ifndef JINJA2_STRING_HELPERS_H
+#define JINJA2_STRING_HELPERS_H
+
+#include <nonstd/string_view.hpp>
+#include "value.h"
+
+#include <string>
+#include <type_traits>
+#include <cwchar>
+#include <iostream>
```
Comment (rmorozov) [inline include/jinja2cpp/string_helpers.h]: remove iostream, prefer fmt::print

File: include/jinja2cpp/string_helpers.h [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                return std::string();
+#endif
+            std::string result;
+#ifndef _MSC_VER
+            result.resize(destBytes + 1);
+            auto converted = std::wcsrtombs(&result[0], &srcPtr, srcSize, &state);
+            if (converted == static_cast<std::size_t>(-1))
+                return std::string();
+            result.resize(converted);
+            // std::wcout << L"Convert '" << from << L"' of " << srcSize << L" chars len to '";
```
Comment (rmorozov) [inline include/jinja2cpp/string_helpers.h]: remove comments?

File: include/jinja2cpp/user_callable.h [hunk 3]
```diff
@@ -15,6 +17,86 @@ struct CanBeCalled : std::false_type {};
 template<typename T>
 struct CanBeCalled<T, typename std::enable_if<std::is_same<typename T::result_type, Value>::value>::type> : std::true_type {};
 
+template<typename T>
+struct ArgPromoter
+{
+    ArgPromoter(const T* val) : m_ptr(val) {}
```
Comment (rmorozov) [inline include/jinja2cpp/user_callable.h]: maybe something like gsl::not_null?

Comment (rmorozov) [reply include/jinja2cpp/user_callable.h]: or at least https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper ?

File: src/filters.cpp [hunk 4]
```diff
@@ -457,64 +460,81 @@ struct PrettyPrinter : visitors::BaseVisitor<InternalValue>
...
+        fmt::format_to(os, "{}", Apply<PrettyPrinter>(kwPair.value, m_context));
 
-        return InternalValue(os.str());
+        return str;
     }
 
-    InternalValue operator()(const std::string& str) const
+    std::string operator()(const std::string& str) const
     {
         return "'"s + str + "'"s;
```
Comment (rmorozov) [inline src/filters.cpp]: fmt::format ?

File: src/filters.cpp [hunk 5]
```diff
@@ -457,64 +460,81 @@ struct PrettyPrinter : visitors::BaseVisitor<InternalValue>
...
     }
 
-    InternalValue operator()(const std::wstring&) const
+    std::string operator()(const nonstd::string_view& str) const
+    {
+        return "'"s + std::string(str.begin(), str.end()) + "'"s;
+    }
+
+    std::string operator()(const std::wstring& str) const
```
Comment (rmorozov) [inline src/filters.cpp]: fmt::format

File: include/jinja2cpp/user_callable.h (hunk 3 above)
Comment (flexferrum) [reply include/jinja2cpp/user_callable.h]: See no reason for that here. It's internal utility class which isn't directly exposed to the user.

File: include/jinja2cpp/filesystem_handler.h [hunk 6]
```diff
@@ -2,6 +2,7 @@
 #define JINJA2CPP_FILESYSTEM_HANDLER_H
 
 #include <nonstd/variant.hpp>
+#include <nonstd/optional.hpp>
 
 #include <iostream>
```
Comment (rmorozov) [inline include/jinja2cpp/filesystem_handler.h]: is iostream needed here?

File: include/jinja2cpp/string_helpers.h [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        {
+            std::mbstate_t state = std::mbstate_t();
+            auto srcPtr = from.data();
+            std::size_t srcSize = from.size();
+            std::size_t destBytes = 0;
+
+#ifndef _MSC_VER
+            destBytes = std::wcsrtombs(nullptr, &srcPtr, srcSize, &state);
+            if (destBytes == static_cast<std::size_t>(-1))
+                return std::string();
```
Comment (rmorozov) [inline include/jinja2cpp/string_helpers.h]: Isn't it better to throw exception here? in such a way we silently swallow error

File: include/jinja2cpp/filesystem_handler.h (hunk 6 above)
Comment (flexferrum) [reply include/jinja2cpp/filesystem_handler.h]: Probably yes, for the declaration of basic_i/ostream etc.

File: include/jinja2cpp/string_helpers.h (hunk 7 above)
Comment (flexferrum) [reply include/jinja2cpp/string_helpers.h]: And what to do with this exception? :) I'll think about it, actually.

**Pull request merged**